### PR TITLE
Preparing Microsoft.Identity.Web for being disributed as a NuGet package

### PR DESCRIPTION
**Pull request by jmprieur**
## Purpose
Preparing Microsoft.Identity.Web for being disributed as a NuGet package
1) Merging the content from microsoft-authentication-extensions-for-dotnet Web
2) Aligning the namespaces with the folder strucuture
3) Renaming files and classes per the style used in MSAL.NET
* ...

## Does this introduce a breaking change?

```
[ x] Yes: 
- the Microsoft.Identity.Web.Client namespace has disapeared (only Microsoft.Identity.Web), and the token cache implementations are now respectively in Microsoft.Identity.Web.Client.TokenCache.Session, Microsoft.Identity.Web.Client.TokenCache.Session.InMemory, Microsoft.Identity.Web.Client.TokenCache.Session.Sql
- The methods of ITokenAcquisition / TokenAcquisition are now Async (renamed)
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

Code clean-up.

```
[ ] Bugfix
[ ] Feature
[ x] Code style update (formatting, local variables)
[ x ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: 2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.Options;
 using Graph=Microsoft.Graph;
-using Microsoft.Identity.Web.Client;
+using Microsoft.Identity.Web;
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs]: In the library, aligned the folder structure and the namespaces (like @MarkZuber had proposed in the extension library)
Note that the library is not only for clients, but also resources / Web APIs (issuer validator for instance) and therefore the root namespace is not Microsoft.Identity.Client, but Microsoft.Identity.Web

File: 2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs [hunk 2]
```diff
@@ -59,7 +59,7 @@ public async Task<IActionResult> Profile()
         {
             return GraphServiceClientFactory.GetAuthenticatedGraphClient(async () =>
             {
-                string result = await tokenAcquisition.GetAccessTokenOnBehalfOfUser(
+                string result = await tokenAcquisition.GetAccessTokenOnBehalfOfUserAsync(
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs]: The methods which used to have a .Wait() are now async

File: 2-WebApp-graph-user/2-1-Call-MSGraph/Startup.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Identity.Web;
-using Microsoft.Identity.Web.Client.TokenCacheProviders;
+using Microsoft.Identity.Web.TokenCacheProviders.InMemory;
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-1-Call-MSGraph/Startup.cs]: The token cache providers have now a sub namespace

File: 2-WebApp-graph-user/2-2-TokenCache/Startup.cs [hunk 4]
```diff
@@ -36,8 +36,8 @@ public void ConfigureServices(IServiceCollection services)
             // and chosen token cache implementation
             services.AddAzureAdV2Authentication(Configuration)
                 .AddMsal(new string[] { Constants.ScopeUserRead })
-                    .AddSqlAppTokenCache(new MSALSqlTokenCacheOptions(Configuration.GetConnectionString("TokenCacheDbConnStr")))
-                    .AddSqlPerUserTokenCache(new MSALSqlTokenCacheOptions(Configuration.GetConnectionString("TokenCacheDbConnStr")));
+                    .AddSqlAppTokenCache(new MsalSqlTokenCacheOptions(Configuration.GetConnectionString("TokenCacheDbConnStr")))
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-2-TokenCache/Startup.cs]: MSAL -> Msal

File: Microsoft.Identity.Web/AccountExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        /// <returns>A <see cref="ClaimsPrincipal"/> built from IAccount</returns>
+        public static ClaimsPrincipal ToClaimsPrincipal(this IAccount account)
+        {
+            if (account != null)
+            {
+                return new ClaimsPrincipal(
+                    new ClaimsIdentity(new Claim[]
+                    {
+                        new Claim(ClaimConstants.Oid, account.HomeAccountId.ObjectId),
+                        new Claim(ClaimConstants.Tid, account.HomeAccountId.TenantId),
```
Comment (bgavrilMS) [inline Microsoft.Identity.Web/AccountExtensions.cs]: Does this work for B2C?

File: Microsoft.Identity.Web.Test/AadIssuerValidatorTests.cs [hunk 6]
```diff
@@ -1,3 +1,6 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jmprieur) [inline Microsoft.Identity.Web.Test/AadIssuerValidatorTests.cs]: Simplified header for the MIT Copyright

File: Microsoft.Identity.Web/AccountExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jmprieur) [inline Microsoft.Identity.Web/AccountExtensions.cs]: Extracted this class in its own file

File: Microsoft.Identity.Web/ClaimsPrincipalFactory.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jmprieur) [inline Microsoft.Identity.Web/ClaimsPrincipalFactory.cs]: Adopted the names that @MarkZuber had proposed

File: Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs [hunk 9]
```diff
@@ -1,183 +1,115 @@
-﻿/************************************************************************************************
```
Comment (jmprieur) [inline Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs]: Adopted the named proposed by @MarkZuber 
Simplified header

File: Microsoft.Identity.Web/Client/TokenCacheProviders/IMSALAppTokenCacheProvider.cs [hunk 10]
```diff
@@ -1,48 +0,0 @@
-﻿/************************************************************************************************
```
Comment (jmprieur) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/IMSALAppTokenCacheProvider.cs]: Renamed to I**Msal**AppTokenCacheProvider

File: Microsoft.Identity.Web/ITokenAcquisition.cs [hunk 11]
```diff
@@ -66,12 +73,14 @@ public interface ITokenAcquisition
...
         /// options.Events.OnTokenValidated = async context =>
         /// {
-        ///   var tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();
+        ///   var tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService&lt;ITokenAcquisition&gt;();
         ///   tokenAcquisition.AddAccountToCacheFromJwt(context);
         /// };
         /// </code>
         /// </example>
-        void AddAccountToCacheFromJwt(AspNetCore.Authentication.JwtBearer.TokenValidatedContext tokenValidationContext, IEnumerable<string> scopes = null);
+        Task AddAccountToCacheFromJwtAsync(
```
Comment (jmprieur) [inline Microsoft.Identity.Web/ITokenAcquisition.cs]: Used the same style as @MarkZuber for methods with many parameters

File: Microsoft.Identity.Web/InstanceDiscovery/IssuerMetadata.cs [hunk 12]
```diff
@@ -32,12 +11,21 @@ namespace Microsoft.Identity.Web.InstanceDiscovery
     /// </summary>
     internal class IssuerMetadata
     {
+        /// <summary>
```
Comment (jmprieur) [inline Microsoft.Identity.Web/InstanceDiscovery/IssuerMetadata.cs]: Added XML comments

File: Microsoft.Identity.Web/Resource/ScopesRequiredByWebAPIExtension.cs [hunk 13]
```diff
@@ -1,36 +1,12 @@
...
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
-***********************************************************************************************/
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
 
 using Microsoft.AspNetCore.Http;
-using Microsoft.AspNetCore.Mvc;
 using System;
-using System.Collections.Generic;
```
Comment (jmprieur) [inline Microsoft.Identity.Web/Resource/ScopesRequiredByWebAPIExtension.cs]: Removed unused usings

File: Microsoft.Identity.Web/AccountExtensions.cs (hunk 5 above)
Comment (jmprieur) [reply Microsoft.Identity.Web/AccountExtensions.cs]: It used to work before, and I have not changed the semantics. This should. Do you have a concern here, @bgavrilMS ?

**Comment by jmprieur** [comment]
Thanks @bgavrilMS 
I re-tested most of the samples in the sub-folders (and fixed 2 regressions I had introduced, BTW)
This is looking good

File: Microsoft.Identity.Web/AccountExtensions.cs (hunk 5 above)
Comment (TiagoBrenck) [reply Microsoft.Identity.Web/AccountExtensions.cs]: @jmprieur I don't if this was Bogdan`s concern, but `TID` is not presented on B2C tokens.

**Comment by deisterhold** [comment]
FYI. I think there is a bug in MsalAppMemoryTokenCacheProvider where `option != null` should be `option == null`.

See #162.

File: Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs [hunk 14]
```diff
@@ -16,15 +19,22 @@ namespace Microsoft.Identity.Web.Client
     /// The following controller action will trigger
     /// <code>
     /// [MsalUiRequiredExceptionFilter(Scopes = new[] {"Mail.Send"})]
-    /// public async Task<IActionResult> SendEmail()
+    /// public async Task&lt;IActionResult&gt; SendEmail()
     /// {
     /// }
     /// </code>
     /// </example>
     public class MsalUiRequiredExceptionFilterAttribute : ExceptionFilterAttribute
```
Comment (kalyankrishna1) [inline Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs]: >MsalUiRequiredExceptionFilterAttribute [](start = 17, length = 38)

Learning from the existing AuthorizeAttribute in use in as.net and asp.net core, the attribute traps the AccessDeniedException and acts on it. This attribute is somewhat similar in that respect. Can we consider renaming this to MsalAuthorizeAttribute or MsalAuthorizeForScopesAttribute or something similar .IMHO, using an exception in an attribute name like we have for  MsalUiRequiredExceptionFilterAttribute, carries somewhat negative connotations.

File: Microsoft.Identity.Web/Resource/ScopesRequiredHttpContextExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        /// <summary>
+        /// When applied to an <see cref="HttpContext"/>, verifies that the user authenticated in the
+        /// web API has any of the accepted scopes. 
+        /// If the authentication user does not have any of these <paramref name="acceptedScopes"/>, the
+        /// method throws an HTTP Unauthorized with the message telling which scopes are expected in the token
+        /// </summary>
+        /// <param name="acceptedScopes">Scopes accepted by this web API</param>
+        /// <exception cref="HttpRequestException"/> with a <see cref="HttpResponse.StatusCode"/> set to 
+        /// <see cref="HttpStatusCode.Unauthorized"/>
+        public static void VerifyUserHasAnyAcceptedScope(this HttpContext context, params string[] acceptedScopes)
```
Comment (kalyankrishna1) [inline Microsoft.Identity.Web/Resource/ScopesRequiredHttpContextExtensions.cs]: >VerifyUserHasAnyAcceptedScope [](start = 27, length = 29)

This pattern requires a call to this method to be placed in each and every one of every controller methods, which is very excessive. A check for needed scopes can be done in middleware, preferably after the token has been validated.

Comment (jmprieur) [reply Microsoft.Identity.Web/Resource/ScopesRequiredHttpContextExtensions.cs]: I disagree on this one, @kalyankrishna1 : you are assuming that the web app / web api will have only one controller / action.  I believe that real life applications have several actions, which would be each protected by their scope (at least the developer should chose)

File: Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs [hunk 16]
```diff
@@ -1,183 +1,115 @@
...
+
+        /// <summary>
+        /// Gets the unique object ID associated with the <see cref="ClaimsPrincipal"/>
+        /// </summary>
+        /// <param name="claimsPrincipal">the <see cref="ClaimsPrincipal"/> from which to retrieve the unique object id</param>
+        /// <returns>Unique object ID of the identity, or <c>null</c> if it cannot be found</returns>
+        public static string GetObjectId(this ClaimsPrincipal claimsPrincipal)
+        {
+            string userObjectId = claimsPrincipal.FindFirstValue(ClaimConstants.Oid);
+            if (string.IsNullOrEmpty(userObjectId))
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs]: Brent changed this. Thanks for confirming that this is the style we use, @MarkZuber

File: Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs (hunk 14 above)
Comment (jmprieur) [reply Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs]: I like MsalAuthorizeForScopesAttribute or AuthorizeForScopesAttribute
What about EnsureScopesAttribute?

File: .gitignore [hunk 17]
```diff
@@ -96,3 +96,5 @@
 /4-WebApp-your-API/TodoListService/obj
 /4-WebApp-your-API/TodoListService/bin
 /4-WebApp-your-API/Client/obj
+/2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph/bin/Debug/netcoreapp2.2
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply .gitignore]: Does it cover /.vs, and all the .vs folders in the sub folders, @MarkZuber ?

File: 4-WebApp-your-API/TodoListService/Startup.cs [hunk 18]
```diff
@@ -44,17 +47,26 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            // This is required to be instantiated before the OpenIdConnectOptions starts getting configured.
+            // By default, the claims mapping will map claim names in the old format to accommodate older SAML applications.
+            // 'http://schemas.microsoft.com/ws/2008/06/identity/claims/role' instead of 'roles'
+            // This flag ensures that the ClaimsIdentity claims collection will be built from the claims in the token
+            JwtSecurityTokenHandler.DefaultMapInboundClaims = false;
+
             services.AddProtectWebApiWithMicrosoftIdentityPlatformV2(Configuration)
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply 4-WebApp-your-API/TodoListService/Startup.cs]: Done https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/commit/df88abbd5430dd0a0d029d62f180b6e0ee5a6c58

File: Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/ClaimsPrincipalExtensions.cs]: Done

File: Microsoft.Identity.Web/ServiceCollectionExtensions.cs [hunk 19]
```diff
@@ -61,13 +39,12 @@ public static IServiceCollection AddTokenAcquisition(this IServiceCollection ser
             services.AddScoped<ITokenAcquisition>(factory =>
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/ServiceCollectionExtensions.cs]: Done. Thanks!

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 20]
```diff
@@ -62,38 +42,44 @@ public class TokenAcquisition : ITokenAcquisition
         /// <param name="appTokenCacheProvider">The App token cache provider</param>
         /// <param name="userTokenCacheProvider">The User token cache provider</param>
         /// <param name="configuration"></param>
-        public TokenAcquisition(IConfiguration configuration, IMSALAppTokenCacheProvider appTokenCacheProvider, IMSALUserTokenCacheProvider userTokenCacheProvider)
+        public TokenAcquisition(
+            IConfiguration configuration,
+            IMsalAppTokenCacheProvider appTokenCacheProvider,
+            IMsalUserTokenCacheProvider userTokenCacheProvider)
         {
             if (configuration == null)
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: Done!

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 21]
```diff
@@ -154,16 +143,19 @@ public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceiv
...
         /// <param name="scopes">Scopes to request for the downstream API to call</param>
-        /// <param name="tenantId">Enables to override the tenant/account for the same identity. This is useful in the
-        /// cases where a given account is guest in other tenants, and you want to acquire tokens for a specific tenant</param>
-        /// <returns>An access token to call on behalf of the user, the downstream API characterized by its scopes</returns>
-        public async Task<string> GetAccessTokenOnBehalfOfUser(HttpContext context, IEnumerable<string> scopes, string tenant = null)
+        /// <param name="tenant">Enables overriding of the tenant/account for the same identity. This is useful in the
+        /// cases where a given account is guest in other tenants, and you want to acquire tokens for a specific tenant, like where the user is a guest in</param>
+        /// <returns>An access token to call the downstream API and populated with this downstream Api's scopes</returns>
+        public async Task<string> GetAccessTokenOnBehalfOfUserAsync(
+            HttpContext context,
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: @MarkZuber : I added it in services.AddTokenAcquisition(). Does it work this way?

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 22]
```diff
@@ -310,29 +308,34 @@ private IConfidentialClientApplication GetOrBuildConfidentialClientApplication(H
...
+
+            var app = ConfidentialClientApplicationBuilder
+                .CreateWithApplicationOptions(_applicationOptions)
+                .WithRedirectUri(currentUri)
+                .WithAuthority(authority)
+                .Build();
 
             // Initialize token cache providers
-            if (this.AppTokenCacheProvider != null)
+            if (_appTokenCacheProvider != null)
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: Done

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 23]
```diff
@@ -395,14 +418,15 @@ private async Task<string> GetAccessTokenOnBehalfOfUser(IConfidentialClientAppli
         /// <summary>
         /// Adds an account to the token cache from a JWT token and other parameters related to the token cache implementation
         /// </summary>
-        private void AddAccountToCacheFromJwt(IEnumerable<string> scopes, JwtSecurityToken jwtToken, ClaimsPrincipal principal, HttpContext httpContext)
+        private async Task AddAccountToCacheFromJwtAsync(IEnumerable<string> scopes, JwtSecurityToken jwtToken, ClaimsPrincipal principal, HttpContext httpContext)
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: I removed the HttpContext
**However**, the ClaimsPrincipal which is passed can be different from the claims principal in the HttpContext (depending on the stage in the middleware pipeline, there are cases where the ClaimsPrincipal in the HttpContext can have less claims than the ClaimsPrincipal to which the developer have access. Also there is also another scenario of (MSGraph) notifications where the developer rebuilds a claims principal based on the oid/tid which are sent in the notification.

File: Microsoft.Identity.Web/TokenCacheProviders/IMSALUserTokenCacheProvider.cs [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the MIT License.
+
+using Microsoft.Identity.Client;
+
+namespace Microsoft.Identity.Web.TokenCacheProviders
+{
+    /// <summary>
+    /// MSAL token cache provider interface for user accounts
+    /// </summary>
+    public interface IMsalUserTokenCacheProvider
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/IMSALUserTokenCacheProvider.cs]: @MarkZuber : I agree
would you want to propose a PR on top of this one ?

File: Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalAppMemoryTokenCacheProvider.cs [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+            _memoryCache = cache;
+        }
+
+        /// <summary>Initializes this instance of TokenCacheProvider with essentials to initialize themselves.</summary>
+        /// <param name="tokenCache">The token cache instance of MSAL application</param>
+        public void Initialize(ITokenCache tokenCache)
+        {
+            _appCacheId = _appId + "_AppTokenCache";
+
+            tokenCache.SetBeforeAccess(AppTokenCacheBeforeAccessNotification);
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalAppMemoryTokenCacheProvider.cs]: yes, now that we have them :)

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs [hunk 26]
```diff
@@ -154,62 +135,38 @@ public void Clear()
...
+            ReadCacheForSignedInApp(args);
         }
 
         /// <summary>
         /// Reads the cache data from the backend database.
         /// </summary>
         private void ReadCacheForSignedInApp(TokenCacheNotificationArgs args)
         {
-            if (this.InMemoryCache == null) // first time access
+            if (_inMemoryCache == null) // first time access
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs]: @kalyankrishna1 : what did you have in mind?

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 27]
```diff
@@ -62,38 +42,44 @@ public class TokenAcquisition : ITokenAcquisition
...
+            IConfiguration configuration,
+            IMsalAppTokenCacheProvider appTokenCacheProvider,
+            IMsalUserTokenCacheProvider userTokenCacheProvider)
         {
             if (configuration == null)
                 throw new ArgumentNullException(nameof(configuration));
 
-            azureAdOptions = new AzureADOptions();
-            configuration.Bind("AzureAD", azureAdOptions);
+            _azureAdOptions = new AzureADOptions();
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: As @TiagoBrenck mentioned, we might want to pass the section name or the delegate as a parameter.

File: Microsoft.Identity.Web/TokenAcquisition.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: @MarkZuber : please ack on the however. How can we make it clear that this scenario exists, BTW?

File: 2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph/Controllers/HomeController.cs [hunk 28]
```diff
@@ -31,7 +31,7 @@ public IActionResult Index()
         public async Task<IActionResult> Profile()
         {
             var accessToken =
-                await tokenAcquisition.GetAccessTokenOnBehalfOfUser(HttpContext, new[] {Constants.ScopeUserRead});
+                await tokenAcquisition.GetAccessTokenOnBehalfOfUserAsync(HttpContext, new[] {Constants.ScopeUserRead});
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply 2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph/Controllers/HomeController.cs]: Done

File: Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalAppMemoryTokenCacheProvider.cs [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+using System;
+
+namespace Microsoft.Identity.Web.TokenCacheProviders.InMemory
+{
+    /// <summary>
+    /// An implementation of token cache for Confidential clients backed by MemoryCache.
+    /// MemoryCache is useful in Api scenarios where there is no HttpContext to cache data.
+    /// </summary>
+    /// <seealso cref="https://aka.ms/msal-net-token-cache-serialization"/>
+    public class MsalAppMemoryTokenCacheProvider : IMsalAppTokenCacheProvider
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalAppMemoryTokenCacheProvider.cs]: Done. Thanks Mark for the suggestion!

File: Microsoft.Identity.Web/WebApiServiceCollectionExtensions.cs [hunk 30]
```diff
@@ -127,7 +124,7 @@ public static IServiceCollection AddProtectedApiCallsWebApis(this IServiceCollec
                     {
                         var tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();
                         context.Success();
-                        tokenAcquisition.AddAccountToCacheFromJwt(context, scopes);
+                        await tokenAcquisition.AddAccountToCacheFromJwtAsync(context, scopes);
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/WebApiServiceCollectionExtensions.cs]: Done

File: Microsoft.Identity.Web/ServiceCollectionExtensions.cs [hunk 31]
```diff
@@ -36,14 +36,8 @@ public static class ServiceCollectionExtensions
...
-            services.AddScoped<ITokenAcquisition>(factory =>
-            {
-                var config = factory.GetRequiredService<IConfiguration>();
-                var apptokencacheprovider = factory.GetService<IMsalAppTokenCacheProvider>();
-                var usertokencacheprovider = factory.GetService<IMsalUserTokenCacheProvider>();
-
-                return new TokenAcquisition(config, apptokencacheprovider, usertokencacheprovider);
-            });
+            services.AddHttpContextAccessor();
+            services.AddScoped<ITokenAcquisition>();
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/ServiceCollectionExtensions.cs]: Done. Thanks!

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 32]
```diff
@@ -33,22 +33,29 @@ public class TokenAcquisition : ITokenAcquisition
         private readonly IMsalUserTokenCacheProvider _userTokenCacheProvider;
 
         private IConfidentialClientApplication application;
+        private IHttpContextAccessor httpContextAccessor;
+        private HttpContext HttpContext { get { return httpContextAccessor.HttpContext; } }
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: Done. Thanks!

File: Microsoft.Identity.Web/TokenAcquisition.cs [hunk 33]
```diff
@@ -147,30 +158,27 @@ public async Task AddAccountToCacheFromAuthorizationCodeAsync(AuthorizationCodeR
...
                 throw new ArgumentNullException(nameof(scopes));
+            }
 
             // Use MSAL to get the right token to call the API
-            var application = GetOrBuildConfidentialClientApplication(context);
+            var application = GetOrBuildConfidentialClientApplication();
 
             // Case of a lazy OBO
-            Claim jwtClaim = context.User.FindFirst("jwt");
+            Claim jwtClaim = HttpContext.User.FindFirst("jwt");
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenAcquisition.cs]: Thanks. I renamed the property to CurrentHttpContext.

File: Microsoft.Identity.Web/TokenCacheProviders/IMSALUserTokenCacheProvider.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/IMSALUserTokenCacheProvider.cs]: Done.

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/AppTokenCache.cs [hunk 34]
```diff
@@ -0,0 +1,41 @@
...
+
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace Microsoft.Identity.Web.TokenCacheProviders.Sql
+{
+    /// <summary>
+    /// Represents an app's token cache entry in database
+    /// </summary>
+    public class AppTokenCache
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/AppTokenCache.cs]: Done.

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalPerUserSqlTokenCacheProvider.cs [hunk 35]
```diff
@@ -1,96 +1,76 @@
...
         /// <param name="tokenCache">The token cache instance of MSAL application</param>
-        /// <param name="httpcontext">The Httpcontext whose Session will be used for caching.This is required by some providers.</param>
-        /// <param name="user">The signed-in user for whom the cache needs to be established. Not needed by all providers.</param>
-        public void Initialize(ITokenCache tokenCache, HttpContext httpcontext, ClaimsPrincipal user)
+        public void Initialize(ITokenCache tokenCache)
         {
-            tokenCache.SetBeforeAccess(this.UserTokenCacheBeforeAccessNotification);
-            tokenCache.SetAfterAccess(this.UserTokenCacheAfterAccessNotification);
-            tokenCache.SetBeforeWrite(this.UserTokenCacheBeforeWriteNotification);
+            tokenCache.SetBeforeAccess(UserTokenCacheBeforeAccessNotification);
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalPerUserSqlTokenCacheProvider.cs]: Done. Used the abstract Token cache

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/TokenCacheDbContext.cs [hunk 36]
```diff
@@ -0,0 +1,32 @@
...
+        /// <param name="options"></param>
+        public TokenCacheDbContext(DbContextOptions<TokenCacheDbContext> options)
+            : base(options)
+        {
+        }
+
+        /// <summary>
+        /// The app token cache table
+        /// </summary>
+        public DbSet<AppTokenCache> AppTokenCache { get; set; }
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/TokenCacheDbContext.cs]: Done.

File: Microsoft.Identity.Web/WebApiServiceCollectionExtensions.cs [hunk 37]
```diff
@@ -40,24 +18,34 @@
...
+        /// This method expects the configuration file will have a section named "AzureAd" with the necessary settings to initialize authentication options.
         /// </summary>
         /// <param name="services">Service collection to which to add this authentication scheme</param>
         /// <param name="configuration">The Configuration object</param>
+        /// <param name="subscribeToJwtBearerMiddlewareDiagnosticsEvents">
+        /// Set to true if you want to debug, or just understand the JwtBearer events.
+        /// </param>
         /// <returns></returns>
-        public static IServiceCollection AddProtectWebApiWithMicrosoftIdentityPlatformV2(this IServiceCollection services, IConfiguration configuration, X509Certificate2 tokenDecryptionCertificate = null)
+        public static IServiceCollection AddProtectWebApiWithMicrosoftIdentityPlatformV2(
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/WebApiServiceCollectionExtensions.cs]: Done.

File: Microsoft.Identity.Web/WebAppServiceCollectionExtensions.cs [hunk 38]
```diff
@@ -1,50 +1,37 @@
...
+        /// This method expects the configuration file will have a section named "AzureAd" with the necessary settings to initialize authentication options.
         /// </summary>
         /// <param name="services">Service collection to which to add this authentication scheme</param>
         /// <param name="configuration">The Configuration object</param>
+        /// <param name="subscribeToOpenIdConnectMiddlewareDiagnosticsEvents">
+        /// Set to true if you want to debug, or just understand the OpenIdConnect events.
+        /// </param>
         /// <returns></returns>
-        public static IServiceCollection AddAzureAdV2Authentication(this IServiceCollection services, IConfiguration configuration)
+        public static IServiceCollection AddAzureAdV2Authentication(
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/WebAppServiceCollectionExtensions.cs]: Done

File: Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs (hunk 14 above)
Comment (jmprieur) [reply Microsoft.Identity.Web/MsalUiRequiredExceptionFilterAttribute.cs]: I renamed to AuthorizeForScopesAttribute.

File: Microsoft.Identity.Web/TokenCacheProviders/InMemory/InMemoryTokenCacheProviderExtension.cs [hunk 39]
```diff
@@ -29,56 +8,69 @@
...
-        public static IServiceCollection AddInMemoryAppTokenCache(this IServiceCollection services, MSALMemoryTokenCacheOptions cacheOptions)
+        /// <param name="cacheOptions">The MSALMemoryTokenCacheOptions allows the caller to set the token cache expiration</param>
+        public static IServiceCollection AddInMemoryAppTokenCache(
+            this IServiceCollection services,
+            MsalMemoryTokenCacheOptions cacheOptions)
         {
             services.AddMemoryCache();
 
-            services.AddSingleton<IMSALAppTokenCacheProvider>(factory =>
+            services.AddSingleton<IMsalAppTokenCacheProvider>(factory =>
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/InMemory/InMemoryTokenCacheProviderExtension.cs]: Done

**Comment by jmprieur** [comment]
@MarkZuber @TiagoBrenck 
I've taken into account all your feedback, except, @TiagoBrenck for B2C (we need to discuss what to do here)

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (kalyankrishna1) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs]: Took me a while to remember, I couldn't believe this myself . This is a carry over from a perf improvement when Sql cache was sort of singleton for the whole app and this updated the cache when MSAL updated the cache. Not needed any more as the cache is scoped in aspnet lifecycle

[the remark this answers is not in the document]
Comment (jmprieur) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs]: @kalyankrishna1 : shall I propose a PR to get rid of it?
cc: @MarkZuber

[the remark this answers is not in the document]
Comment (kalyankrishna1) [reply Microsoft.Identity.Web/TokenCacheProviders/Sql/MsalAppSqlTokenCacheProvider.cs]: I already have a PR going, expect it today

**Pull request merged**